src/devices/machine/mc6844.cpp

```diff
@@ -0,0 +1,556 @@
+  Aaron Giles.
+
+  TODO:
+  - Memory to Device transfers
+
+**********************************************************************/
+
+#include "emu.h"
+#include "mc6844.h"
+
+//**************************************************************************
+//  GLOBAL VARIABLES
+//**************************************************************************
+#define LOG_SETUP   (1U << 1)
+#define LOG_INT     (1U << 2)
+#define LOG_STATE   (1U << 3)
+#define LOG_TFR     (1U << 4)
+
+//#define VERBOSE (LOG_GENERAL|LOG_SETUP | LOG_INT | LOG_STATE |LOG_TFR)
+//#define LOG_OUTPUT_STREAM std::cout
+
+#include "logmacro.h"
+
+#define LOGSETUP(...) LOGMASKED(LOG_SETUP, __VA_ARGS__)
+#define LOGINT(...)   LOGMASKED(LOG_INT,   __VA_ARGS__)
+#define LOGSTATE(...) LOGMASKED(LOG_STATE, __VA_ARGS__)
+#define LOGTFR(...)   LOGMASKED(LOG_TFR,   __VA_ARGS__)
+
+#ifdef _MSC_VER
+#define FUNCNAME __func__
+#else
+#define FUNCNAME __PRETTY_FUNCTION__
+#endif
+
+// device type definition
+DEFINE_DEVICE_TYPE(MC6844, mc6844_device, "mc6844", "MC6844 DMA")
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  mc6844_device - constructor
+//-------------------------------------------------
+mc6844_device::mc6844_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, MC6844, tag, owner, clock)
+	, device_execute_interface(mconfig, *this)
+	, m_out_int_cb(*this)
+	, m_out_txak_cb(*this)
+	, m_out_drq1_cb(*this)
+	, m_out_drq2_cb(*this)
+	, m_in_memr_cb(*this)
+	, m_out_memw_cb(*this)
+	, m_in_ior_cb{ { *this },{ *this },{ *this },{ *this } }
+	, m_out_iow_cb{ { *this },{ *this },{ *this },{ *this } }
+	, m_state(STATE_S0)
+	, m_icount(0)
+{
+}
+
+//-------------------------------------------------
+//  device_resolve_objects - device-specific setup
+//-------------------------------------------------
+void mc6844_device::device_resolve_objects()
+{
+	m_out_int_cb.resolve_safe();
+	m_out_txak_cb.resolve_safe();
+	m_out_drq1_cb.resolve_safe();
+	m_out_drq2_cb.resolve_safe();
+	m_in_memr_cb.resolve_safe(0);
+	m_out_memw_cb.resolve_safe();
+
+	for(auto &cb : m_in_ior_cb)
+			cb.resolve_safe(0);
+	for(auto &cb : m_out_iow_cb)
+			cb.resolve_safe();
+
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void mc6844_device::device_start()
+{
+	// set our instruction counter
+	set_icountptr(m_icount);
+	
+	save_item(NAME(m_m6844_priority));
+	save_item(NAME(m_m6844_interrupt));
+	save_item(NAME(m_m6844_chain));
+	save_item(NAME(m_state));
+	save_item(NAME(m_icount));
+	save_item(NAME(m_current_channel));
+	save_item(NAME(m_last_channel));
+	save_item(NAME(m_dgrnt));
```

Comment:
Don’t the members of `m_6844_channel` also need saving?

Reply:
Yep, haven't checked out the new struct save yet, will do